### ISSUE TITLE
Type mismatch on location error when compiling HLSL tessellation shaders

### ISSUE THREAD

**Issue by DiligentGraphics** [KhronosGroup/glslang#1738]
I am seeing the following error when creating a PSO with tessellation shaders compiled from HLSL:

```
Diligent Engine: ERROR: Vulkan debug message (validation): UNASSIGNED-CoreValidation-Shader-InterfaceTypeMismatch
                 Type mismatch on location 0.0: 'ptr to output vec3 of float32' vs 'ptr to input arr[1] of struct of (vec3 of float32)'
                 Object[0] (shader module): Handle 0x2b1e, Name: '000002B221293918'
```

Vertex shader:
```hlsl
struct VSIn
{
	uint uiVertexId : SV_VertexID;
};

struct VSOutput
{
	float4 f4Position : SV_Position;
	float3 f3Color	: COLOR;
};

void main(VSIn In,
          out VSOutput Out)
{
    float4 Positions[2];
    Positions[0] = float4(-0.65,0.4,0.0,1.0);
    Positions[1] = float4(-0.65,0.65,0.0,1.0);
    float3 Color[2];
    Color[0] = float3(1.0,0.0,0.0);
    Color[1] = float3(0.0,1.0,0.0);
	Out.f4Position = Positions[In.uiVertexId];
    Out.f3Color = Color[In.uiVertexId];
}
```

Hull shader:
```hlsl
struct VSOutput
{
	float4 f4Position : SV_Position;
	float3 f3Color	: COLOR;
};

// Output patch constant data.
struct HS_CONSTANT_DATA_OUTPUT
{
    float Edges[4]  : SV_TessFactor;
};

HS_CONSTANT_DATA_OUTPUT ConstantHS( InputPatch<VSOutput, 1> p, 
                                    uint BlockID : SV_PrimitiveID,
                                    out float Inside[2] : SV_InsideTessFactor)
{
    HS_CONSTANT_DATA_OUTPUT Factors;
    Factors.Edges[0] = 2.5 + float(BlockID);
    Factors.Edges[1] = 4.8 + frac(p[0].f4Position.x*223.145);
    Factors.Edges[2] = 5.9 + frac(p[0].f4Position.y*142.512);
    Factors.Edges[3] = 7.5 - float(BlockID);
    Inside[0] = 6.2 + p[0].f4Position.x*3.4;
    Inside[1] = 7.3 + p[0].f4Position.y*1.4;
    return Factors;
}

struct HSOutput
{
    float4 Position : POS;
    float3 Color : COL;
};

[domain("quad")]
[partitioning("fractional_even")]
[outputtopology("triangle_cw")]
[outputcontrolpoints(1)]
[patchconstantfunc("ConstantHS")]
[maxtessfactor( (float)(32.f+2.f) )]
HSOutput main(InputPatch<VSOutput, 1> inputPatch, uint uCPID : SV_OutputControlPointID)
{
    HSOutput Out;
    Out.Position = inputPatch[uCPID].f4Position;
    Out.Color = inputPatch[uCPID].f3Color;
    return Out;
}
```

Domain shader:
```hlsl
struct DSOutput
{
	float4 f4Position : SV_Position;
	float3 f3Color	: COLOR;
};


struct HS_CONSTANT_DATA_OUTPUT
{
    float Edges[4]  : SV_TessFactor;
    float Inside[2] : SV_InsideTessFactor;
};

struct HSOutput
{
    float4 Position : POS;
    float3 Color : COL;
};

[domain("quad")]
/* partitioning = fractional_even  , 
   outputtopology = triangle_cw */
void main( HS_CONSTANT_DATA_OUTPUT Input, 
           float2 QuadUV : SV_DomainLocation, 
           OutputPatch<HSOutput, 1> QuadPatch,
           out DSOutput Out)
{
    Out.f4Position.xy = QuadPatch[0].Position.xy + QuadUV.xy * 0.2;
    Out.f4Position.zw = QuadPatch[0].Position.zw;
    Out.f3Color = frac( (QuadPatch[0].Color + float3(QuadUV.xy, 1.0)) * float3(4121.35f, 8232.25f, 9124.42f) );
}
```

Pixel shader:
```hlsl
struct DSOutput
{
	float4 f4Position : SV_Position;
	float3 f3Color	: COLOR;
};
 
void main(DSOutput In,
          out float4 Color : SV_Target)
{
	Color = float4(In.f3Color,1.0);
}
```

Despite the error, shaders seem to work properly. They also work correctly in D3D11 and D3D12 back-ends.

glslang version: https://github.com/KhronosGroup/glslang/commit/5efb004d59601711cdf328c8a8bfbe7f333dc7a0
SPIRV-Tools version: https://github.com/KhronosGroup/SPIRV-Tools/commit/2d52cbee49019da5d664bb7e243fa6339c983c27

**Comment by danginsburg** [comment]
Please check to see if this is fixed with #2217

**Comment by TheMostDiligent** [comment]
I confirm the change fixes the original problem, please close the issue.
